logger.go

```diff
@@ -178,11 +180,57 @@ func (log *Logger) With(fields ...Field) *Logger {
 	if len(fields) == 0 {
 		return log
 	}
+	if log.deferWith {
+		log.core = &deferredWithCore{Core: log.core, fields: fields}
+		return log
+	}
 	l := log.clone()
 	l.core = l.core.With(fields)
 	return l
 }
 
+// todo move this elsewhree
+
+/* deferred with core is a zap core that
+ * defers the evaluation of the core's with
+ * until the first log is written.
+ *
+ * This saves significant memory allocation
+ * and CPU with complex cores
+ */
+
+type deferredWithCore struct {
+	zapcore.Core
```

Comment:
put the sync.Once object adjacent to Core to increase the likelihood it ends up on the same cache line